### ISSUE TITLE
Infinite loop when printing CSV data in eager execution mode 

### ISSUE THREAD

**Issue by magnusart** [tensorflow/tensorflow#23785]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
Mac OS X Mojave
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
No
- TensorFlow installed from (source or binary):
binary via pip (1.12.0)
- TensorFlow version (use command below):
v1.12.0-rc2-3-ga6d8ffae09 1.12.0
- Python version:
Python 3.6.5
- Bazel version (if compiling from source):
N/A
- GCC/Compiler version (if compiling from source):
N/A
- CUDA/cuDNN version:
N/A
- GPU model and memory:
N/A

You can collect some of this information using our environment capture [script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with
python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

**Describe the current behavior**
Following the docs at: https://www.tensorflow.org/tutorials/eager/eager_basics
When using `ds_file = tf.data.TextLineDataset(filename)` the docs works as expected, but when using `tf.data.experimental.make_csv_dataset` function then then the first statement in the print loop is repeated infinitely.

Example output
```
1.12.0

Elements in ds_file:
OrderedDict([('Line 1', ), ('1', )])
OrderedDict([('Line 1', ), ('1', )])
OrderedDict([('Line 1', ), ('1', )])
OrderedDict([('Line 1', ), ('1', )])
OrderedDict([('Line 1', ), ('1', )])
OrderedDict([('Line 1', ), ('1', )])
OrderedDict([('Line 1', ), ('1', )])
OrderedDict([('Line 1', ), ('1', )])
... etc
```

**Describe the expected behavior**
I expect the data API to allow me exploring data in eager execution mode even if I use the CsvDataset class. I can't use `TextLineDataset` as my CSV file contain multi line values in quotation marks. 

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.
```python
import tempfile
import tensorflow as tf

print(tf.__version__)
tf.enable_eager_execution()
_, filename = tempfile.mkstemp()

with open(filename, 'w') as f:
  f.write(""""Line 1",1
"Line 2",2
"Line 3",3""")
#ds_file = tf.data.TextLineDataset(filename)
ds_file = tf.data.experimental.make_csv_dataset(filename, 1)

print('\nElements in ds_file:')
for x in ds_file:
  print(x)
```
**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.
See above.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yongtang: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by feihugis** [comment]
This issue is caused by the parameter `num_epochs` for `make_csv_dataset ()`. The default value for `num_epochs` is `None`, which will make the dataset cycle forever as stated [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/experimental/ops/readers.py#L383-L384). One solution is to specify the value of `num_epochs` as following:
`ds_file = tf.data.experimental.make_csv_dataset(filename, 1, num_epochs=1)`.

Users may easily forget to set `num_epochs` when using `make_csv_dataset ()`. However, the default value of `num_epochs` will cause the infinite loop and break the following pipelines if users don't recoginize it. It may be better to remove the default value of `num_epochs` and forces users to specify a number for it?

@mrry @jsimsa @yongtang Do you have any suggestions/comments?

**Comment by mrry** [comment]
Changing `num_epochs` to default to 1 makes sense to me, but it would break anyone using the API in Estimator input functions that expect an infinitely repeated stream of inputs.

**Comment by magnusart** [comment]
So I made the suggested change and I'm still trying to replace `tf.data.TextLineDataset(filename)` with `tf.data.experimental.make_csv_dataset(...)` as before. 

My modified example is not looping forever now, but it still doesn't work as expected in relation to the documentation.

```python
import tempfile
import tensorflow as tf

print(tf.__version__)
tf.enable_eager_execution()
_, filename = tempfile.mkstemp()

with open(filename, 'w') as f:
  f.write(""""Line 1",1
"Line 2",2
"Line 3",3""")
#ds_file = tf.data.TextLineDataset(filename)
ds_file = tf.data.experimental.make_csv_dataset(
    filename, batch_size=1, num_epochs=1)

print('\nElements in ds_file:')
for x in ds_file:
  print(x)

```

Prints out below.  

Again my expectation based on the example I'm trying to modify would be

1. To get three objects printed `OrderedDict([('Line 1', ), ('1', )])` `OrderedDict([('Line 2', ), ('2', )])` etc. 
2. Not have the same line printed twice. If I change the batch_size argument to something > 1 the behavior changes to only print a single line.

```
1.12.0
tf.Tensor(-1, shape=(), dtype=int64)
{'_buffer_size': ,
 '_input_dataset': }

Elements in ds_file:
OrderedDict([('Line 1', ), ('1', )])
OrderedDict([('Line 1', ), ('1', )])
```

**Comment by jsimsa** [comment]
When I run your program, I get:

```
Elements in ds_file:
OrderedDict([('Line 1', <tf.Tensor: id=210, shape=(1,), dtype=string, numpy=array(['Line 3'], dtype=object)>), ('1', <tf.Tensor: id=209, shape=(1,), dtype=int32, numpy=array([3], dtype=int32)>)])
OrderedDict([('Line 1', <tf.Tensor: id=214, shape=(1,), dtype=string, numpy=array(['Line 2'], dtype=object)>), ('1', <tf.Tensor: id=213, shape=(1,), dtype=int32, numpy=array([2], dtype=int32)>)])
```

This indicates that the first line of the CSV file is interpreted as column names and so you only get two results.

Indeed, if you read the documentation for [tf.data.experimental.make_csv_dataset](https://www.tensorflow.org/api_docs/python/tf/data/experimental/make_csv_dataset), you will find this behavior documented -- in particular, see the `header` and `column_names` arguments. If I add `header=False` and `column_names=["string", "number"]` as arguments to the `make_csv_dataset` call in your program, I get the following input:

```
Elements in ds_file:
OrderedDict([('string', <tf.Tensor: id=322, shape=(1,), dtype=string, numpy=array(['Line 1'], dtype=object)>), ('number', <tf.Tensor: id=321, shape=(1,), dtype=int32, numpy=array([1], dtype=int32)>)])
OrderedDict([('string', <tf.Tensor: id=326, shape=(1,), dtype=string, numpy=array(['Line 3'], dtype=object)>), ('number', <tf.Tensor: id=325, shape=(1,), dtype=int32, numpy=array([3], dtype=int32)>)])
OrderedDict([('string', <tf.Tensor: id=330, shape=(1,), dtype=string, numpy=array(['Line 2'], dtype=object)>), ('number', <tf.Tensor: id=329, shape=(1,), dtype=int32, numpy=array([2], dtype=int32)>)])
```

**Comment by magnusart** [comment]
Thanks for trying it out. Interesting enough, it seems to be behaving differently for you. Are you using the same version as I outlined in my initial post? 
 
As you saw I got the same rows output twice (that was not a typo) and I still get that when I try again. 

Good input about the `column_names`. I went ahead and added the `column_names` argument as you suggested. I now get:
```bash
1.12.0
v1.12.0-rc2-3-ga6d8ffae09 1.12.0

Elements in ds_file:
OrderedDict([('string', ), ('number', )])
OrderedDict([('string', ), ('number', )])
```

**Comment by jsimsa** [comment]
Did you also set `header=False`?

**Comment by magnusart** [comment]
You're right, I missed that argument. I now get three lines. Still it prints the same output `OrderedDict([('string', ), ('number', )])` and not the data of each line.

This issue has been the same throughout: It seems that I only ever get the first line when as output. Regardless if it is an infinite cycle of the dataset or if the dataset is only iterated a single time I only.

Just to be clear: my intention is not to make this into a support case. I just want to be assured that this behavior is not a bug before I spend more time. 

I will then leave you guys to figure out wether to use `None` as a default value or not. :)

@jsimsa you do seem to get the expected behavior when printing. What output does below code get you?

```python
import tempfile
import tensorflow as tf

print(tf.__version__)
print(tf.GIT_VERSION, tf.VERSION)

tf.enable_eager_execution()
_, filename = tempfile.mkstemp()

with open(filename, 'w') as f:
  f.write(""""Line 1",1
"Line 2",2
"Line 3",3""")
#ds_file = tf.data.TextLineDataset(filename)
ds_file = tf.data.experimental.make_csv_dataset(
    filename, 
    batch_size=1, 
    num_epochs=1,
    header=False,
    column_names=["string", "number"]
  )

print('\nElements in ds_file:')
for x in ds_file:
  print(x)
```

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.

**Comment by mohantym** [comment]
Hi @magnusart!  your last edited code is not replicate infinite loop issue neither in 1.13 or 2.6 , attached [gist](https://colab.research.google.com/gist/mohantym/7175c5de19129389fd8d0fb3389c16e8/github_23785.ipynb) for reference.
